File: src/lib/arena/prompt-loader.ts
/**
 * Loads AI prompts from Supabase `arena_prompts` table.
 * Falls back to hardcoded defaults if Supabase is unreachable.
 * In-memory cache with 5-minute TTL to avoid per-request DB hits.
 */

import { supabaseAdmin } from '../supabase-admin';

interface CacheEntry {
  content: string;
  fetchedAt: number;
}

const CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes
const cache = new Map<string, CacheEntry>();

/**
 * Load a prompt by ID from Supabase (cached).
 * Returns the prompt content string, or null if not found.
 */
export async function loadPrompt(id: string): Promise<string | null> {
  // Check cache
  const cached = cache.get(id);
  if (cached && Date.now() - cached.fetchedAt < CACHE_TTL_MS) {
    return cached.content;
  }

  try {
    const { data, error } = await supabaseAdmin
      .from('arena_prompts')
      .select('content')
      .eq('id', id)
      .eq('is_active', true)
      .single();

    if (error || !data?.content) {
      console.warn(`[prompt-loader] No active prompt for "${id}":`, error?.message);
      return null;
    }

    cache.set(id, { content: data.content, fetchedAt: Date.now() });
    return data.content;
  } catch (err) {
    console.error(`[prompt-loader] Failed to load prompt "${id}":`, err);
    return null;
  }
}

/** Invalidate a specific cached prompt (useful after edits). */
export function invalidatePromptCache(id?: string) {
  if (id) {
    cache.delete(id);
  } else {
    cache.clear();
  }
}

File: scripts/arena_analysis/prompt_loader.py
"""
Loads AI prompts from Supabase `arena_prompts` table.
Falls back to hardcoded defaults if Supabase is unreachable.
No cache — always fetches fresh so prompt changes take effect immediately.
"""
from __future__ import annotations

import logging

logger = logging.getLogger(__name__)


async def load_prompt(prompt_id: str, fallback: str | None = None) -> str | None:
    """
    Load a prompt by ID from Supabase (always fresh).
    Returns the prompt content string, or the fallback if not found.
    """
    try:
        from arena_analysis.config import settings
        from supabase import create_client

        client = create_client(settings.supabase_url, settings.supabase_key)
        result = (
            client.table("arena_prompts")
            .select("content")
            .eq("id", prompt_id)
            .eq("is_active", True)
            .single()
            .execute()
        )

        if result.data and result.data.get("content"):
            content = result.data["content"]
            print(f"[PromptLoader] ✓ Loaded '{prompt_id}' from Supabase ({len(content)} chars)", flush=True)
            return content

        print(f"[PromptLoader] ✗ No active prompt for '{prompt_id}', using fallback", flush=True)
        return fallback
    except Exception as exc:
        print(f"[PromptLoader] ✗ Failed to load '{prompt_id}': {exc} — using fallback", flush=True)
        return fallback
